verification/src/http_server/handlers/verification/sourcify/api.rs

```diff
@@ -0,0 +1,119 @@
+use crate::{VerificationResponse, VerificationResult};
+use actix_web::{error, error::Error};
+use futures::Future;
+use reqwest::Url;
+use std::sync::Arc;
+
+use super::types::{ApiFilesResponse, ApiRequest, ApiVerificationResponse, Files};
+
+#[async_trait::async_trait]
+pub(super) trait SourcifyApi {
+    async fn verification_request(
+        &self,
+        params: &ApiRequest,
+    ) -> Result<ApiVerificationResponse, reqwest::Error>;
+
+    async fn source_files_request(
+        &self,
+        params: &ApiRequest,
+    ) -> Result<ApiFilesResponse, reqwest::Error>;
+}
+
+pub struct SourcifyApiClient {
+    host: Url,
+    request_timeout: u64,
+    verification_attempts: u64,
+}
+
+impl SourcifyApiClient {
+    pub fn new(host: Url, request_timeout: u64, verification_attempts: u64) -> Self {
+        Self {
+            host,
+            request_timeout,
+            verification_attempts,
+        }
+    }
+
+    async fn make_retrying_request<F, Fut, Response>(
+        &self,
+        request: F,
+    ) -> Result<Response, reqwest::Error>
+    where
+        F: Fn() -> Fut,
+        Fut: Future<Output = Result<Response, reqwest::Error>>,
```

Comment:
can we make the output be generic parameter too? like with `tokio::spawn`

Reply:
ah, we can't, because we need to check for is_err. let's make the error type generic then